app/controllers/incomeSources/add/BusinessAddedObligationsController.scala

```diff
@@ -0,0 +1,167 @@
+
+class BusinessAddedObligationsController @Inject()(authenticate: AuthenticationPredicate,
+                                                   val authorisedFunctions: AuthorisedFunctions,
+                                                   checkSessionTimeout: SessionTimeoutPredicate,
+                                                   retrieveNino: NinoPredicate,
+                                                   val retrieveIncomeSources: IncomeSourceDetailsPredicate,
+                                                   val retrieveBtaNavBar: NavBarPredicate,
+                                                   val itvcErrorHandler: ItvcErrorHandler,
+                                                   incomeSourceDetailsService: IncomeSourceDetailsService,
+                                                   nextUpdatesService: NextUpdatesService,
+                                                   val obligationsView: BusinessAddedObligations)
+                                                  (implicit val appConfig: FrontendAppConfig,
+                                                   implicit val itvcErrorHandlerAgent: AgentItvcErrorHandler,
+                                                   implicit override val mcc: MessagesControllerComponents,
+                                                   val ec: ExecutionContext,
+                                                   dateService: DateServiceInterface)
+  extends ClientConfirmedController with I18nSupport with FeatureSwitching{
+
+  lazy val backUrl: String = controllers.incomeSources.add.routes.AddBusinessReportingMethod.show().url
+  private lazy val agentBackUrl = controllers.incomeSources.add.routes.AddBusinessReportingMethod.showAgent().url
+
+  def show(id: Option[String]): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(isAgent = false, id)
+  }
+
+  def showAgent(id: Option[String]): Action[AnyContent] =
+    Authenticated.async {
+      implicit request =>
+        implicit user =>
+          getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
+            implicit mtdItUser =>
+              handleRequest(isAgent = true, id)
+          }
+    }
+
+  def handleRequest(isAgent: Boolean, id: Option[String])(implicit user: MtdItUser[_], ec: ExecutionContext): Future[Result] = {
+    if (isDisabled(IncomeSources)) {
+      Future.successful(Redirect(controllers.routes.HomeController.show()))
+    } else {
+      id match {
+        case None => Logger("application").error(
+          s"[BusinessAddedObligationsController][handleRequest] - Error: No id supplied by reporting method page")
+          Future(itvcErrorHandler.showInternalServerError())
+        case Some(value) =>
+          val addedBusiness: BusinessDetailsModel = user.incomeSources.businesses.find(x => x.incomeSourceId.contains(value)).get
```

Comment:
this would cause error in case in case businessModel no provided
this would be much better to use for-comprehension in this situation, like:
`
maybeId :  Option[String] // in the method definition
.....
      for {
        incomeSourceId <- maybeId
        addedBusiness <- user.incomeSources.businesses.find(x => x.incomeSourceId.contains(incomeSourceId))
      } yield addedBusiness`

Reply:
I had a case for if the business with the id did not exist but in the above comment you said that was a "redundant step" so I removed it. Will implement something like this now